From 611fd2aeeb1f6412f98c1e8de2578921c9e98bed Mon Sep 17 00:00:00 2001
From: Saurav Shah <quic_sauravs@quicinc.com>
Date: Wed, 12 Jun 2024 11:28:06 +0530
Subject: [PATCH] qdl: Add support for pcie channel

Add support for pcie channel.

This patch file is from Qualcomm Innovation Center, Inc. and is provided under the following license:

Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
SPDX-License-Identifier: BSD-3-Clause-Clear 

Signed-off-by: Saurav Shah <quic_sauravs@quicinc.com>
---
 0001-qdl-Change-usb-macro.patch | 29 ------------
 Makefile                        |  4 +-
 pcie.c                          | 82 +++++++++++++++++++++++++++++++++
 qdl.c                           | 32 +++++++++++++
 usb.c                           |  6 +--
 5 files changed, 119 insertions(+), 34 deletions(-)
 delete mode 100644 0001-qdl-Change-usb-macro.patch
 create mode 100644 pcie.c

diff --git a/0001-qdl-Change-usb-macro.patch b/0001-qdl-Change-usb-macro.patch
deleted file mode 100644
index 1f4a6e7..0000000
--- a/0001-qdl-Change-usb-macro.patch
+++ /dev/null
@@ -1,29 +0,0 @@
-From 9a3888b7013d5cca251b0a63585fd211362e9eac Mon Sep 17 00:00:00 2001
-From: Saurav Shah <sauravs@qti.qualcomm.com>
-Date: Mon, 10 Jun 2024 20:46:59 +0530
-Subject: [PATCH] qdl: Change usb macro
-
-Change usb macro to fix compilation
-issues.
-
-Signed-off-by: Saurav Shah <sauravs@qti.qualcomm.com>
----
- usb.c | 2 +-
- 1 file changed, 1 insertion(+), 1 deletion(-)
-
-diff --git a/usb.c b/usb.c
-index 35a1c1e..fbdce7a 100644
---- a/usb.c
-+++ b/usb.c
-@@ -81,7 +81,7 @@ static int qdl_try_open(libusb_device *dev, struct qdl_device *qdl, const char *
- 			endpoint = &ifc->endpoint[l];
- 
- 			type = endpoint->bmAttributes & LIBUSB_TRANSFER_TYPE_MASK;
--			if (type != LIBUSB_ENDPOINT_TRANSFER_TYPE_BULK)
-+			if (type != LIBUSB_TRANSFER_TYPE_BULK)
- 				continue;
- 
- 			if (endpoint->bEndpointAddress & LIBUSB_ENDPOINT_IN) {
--- 
-2.17.1
-
diff --git a/Makefile b/Makefile
index 4869687..b9b6b11 100644
--- a/Makefile
+++ b/Makefile
@@ -5,10 +5,10 @@ CFLAGS += -O2 -Wall -g `pkg-config --cflags libxml-2.0 libusb-1.0`
 LDFLAGS += `pkg-config --libs libxml-2.0 libusb-1.0`
 prefix := /usr/local
 
-QDL_SRCS := firehose.c qdl.c sahara.c util.c patch.c program.c ufs.c usb.c
+QDL_SRCS := firehose.c qdl.c sahara.c util.c patch.c program.c ufs.c usb.c pcie.c
 QDL_OBJS := $(QDL_SRCS:.c=.o)
 
-RAMDUMP_SRCS := ramdump.c sahara.c usb.c util.c
+RAMDUMP_SRCS := ramdump.c sahara.c usb.c util.c pcie.c
 RAMDUMP_OBJS := $(RAMDUMP_SRCS:.c=.o)
 
 KS_OUT := ks
diff --git a/pcie.c b/pcie.c
new file mode 100644
index 0000000..3c02268
--- /dev/null
+++ b/pcie.c
@@ -0,0 +1,82 @@
+#include <sys/ioctl.h>
+#include <sys/types.h>
+#include <err.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <errno.h>
+#include <poll.h>
+
+#include "qdl.h"
+
+int qdl_open_pcie(struct qdl_device *qdl, const char *serial)
+{
+    int ret;
+	if(NULL == serial)
+    {
+        warnx("serial port is not specified");
+        return -1;
+    }
+    ret = open(serial, O_RDWR | O_SYNC);
+    if(ret != 0)
+    {
+        warnx("Failed to open port (%d)", errno);
+        return -1;
+    }
+    qdl->fd = ret;
+    return 0;
+}
+
+int qdl_read_pcie(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout)
+{
+    int ret;
+    struct pollfd fds[] = {{qdl->fd, POLLIN, 0}};
+
+    ret = poll(fds, 1, timeout);
+    if(ret <= 0)
+    {
+        warnx("qdl_read_pcie error : %d", ret);
+        return -1;
+    }
+    else if(!(fds[0].revents & POLLIN))
+    {
+        warnx("qdl_read_pcie revents error : %d", fds[0].revents);
+        return -1;
+    }
+	return read(qdl->fd, buf, len);
+}
+
+int qdl_write_pcie(struct qdl_device *qdl, const void *buf, size_t len)
+{
+	int ret;
+    struct pollfd fds[] = {{qdl->fd, POLLOUT, 0}};
+    unsigned char *data = (unsigned char*) buf;
+	unsigned int count = 0;
+
+    ret = poll(fds, 1, 1000);
+    if(ret <= 0)
+    {
+        warnx("qdl_write_pcie error : %d", ret);
+        return -1;
+    }
+    else if(!(fds[0].revents & POLLOUT))
+    {
+        warnx("qdl_write_pcie revents error : %d", fds[0].revents);
+        return -1;
+    }
+    while (len > 0) {
+        ret = write(qdl->fd, data, len);
+        if(ret < 0)
+        {
+            warnx("qdl_write_pcie write failed : %d", ret);
+            return -1;
+        }
+        count += ret;
+        len -= ret;
+        data += ret;
+    }
+	return count;
+}
diff --git a/qdl.c b/qdl.c
index 9070cf0..d6c5f3a 100644
--- a/qdl.c
+++ b/qdl.c
@@ -55,6 +55,23 @@ enum {
 bool qdl_debug;
 static struct qdl_device qdl;
 
+
+int (*qdl_open_ptr)(struct qdl_device *qdl, const char *serial);
+int (*qdl_read_ptr)(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout);
+int (*qdl_write_ptr)(struct qdl_device *qdl, const void *buf, size_t len);
+
+inline int qdl_open(struct qdl_device *qdl, const char *serial){return (*qdl_open_ptr)(qdl, serial);}
+inline int qdl_read(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout){return (*qdl_read_ptr)(qdl, buf, len, timeout);}
+inline int qdl_write(struct qdl_device *qdl, const void *buf, size_t len){return (*qdl_write_ptr)(qdl, buf, len);}
+
+extern int qdl_open_usb(struct qdl_device *qdl, const char *serial);
+extern int qdl_read_usb(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout);
+extern int qdl_write_usb(struct qdl_device *qdl, const void *buf, size_t len);
+
+extern int qdl_open_pcie(struct qdl_device *qdl, const char *serial);
+extern int qdl_read_pcie(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout);
+extern int qdl_write_pcie(struct qdl_device *qdl, const void *buf, size_t len);
+
 static int detect_type(const char *xml_file)
 {
 	xmlNode *root;
@@ -154,6 +171,21 @@ int main(int argc, char **argv)
 		return 1;
 	}
 
+	//set the handlers
+	if(strcmp(channel, "pcie") != 0)
+	{
+		qdl_open_ptr = &qdl_open_usb;
+		qdl_read_ptr = &qdl_read_usb;
+		qdl_write_ptr = &qdl_write_usb;
+	}
+	else
+	{
+		qdl_open_ptr = &qdl_open_pcie;
+		qdl_read_ptr = &qdl_read_pcie;
+		qdl_write_ptr = &qdl_write_pcie;
+	}
+	
+
 	if(strcmp(channel, "pcie") != 0)
 	{
 		prog_mbn = argv[optind++];
diff --git a/usb.c b/usb.c
index fbdce7a..ccddf88 100644
--- a/usb.c
+++ b/usb.c
@@ -135,7 +135,7 @@ static int qdl_try_open(libusb_device *dev, struct qdl_device *qdl, const char *
 	return 0;
 }
 
-int qdl_open(struct qdl_device *qdl, const char *serial)
+int qdl_open_usb(struct qdl_device *qdl, const char *serial)
 {
 	struct libusb_device **devs;
 	struct libusb_device *dev;
@@ -180,7 +180,7 @@ int qdl_open(struct qdl_device *qdl, const char *serial)
 	return -1;
 }
 
-int qdl_read(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout)
+int qdl_read_usb(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout)
 {
 	int actual;
 	int ret;
@@ -192,7 +192,7 @@ int qdl_read(struct qdl_device *qdl, void *buf, size_t len, unsigned int timeout
 	return actual;
 }
 
-int qdl_write(struct qdl_device *qdl, const void *buf, size_t len)
+int qdl_write_usb(struct qdl_device *qdl, const void *buf, size_t len)
 {
 	unsigned char *data = (unsigned char*) buf;
 	unsigned int count = 0;
-- 
2.17.1

